cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java

```diff
@@ -887,16 +885,77 @@ public void testCrossNSDatasetAccessWithAuthFlowlet() throws Exception {
     // switch to BOB
     SecurityRequestContext.setUserId(BOB.getName());
 
-    // Send data to stream as BOB this ensures that BOB can do stuff in auth namespace
+    // Send data to stream as BOB this ensures that BOB can write to a stream in AUTH namespace
     StreamManager streamManager = getStreamManager(AUTH_NAMESPACE.toId(), CrossNsDatasetAccessApp.STREAM_NAME);
     for (int i = 0; i < 10; i++) {
       streamManager.send(String.valueOf(i).getBytes());
     }
 
+    // switch to back to ALICE
+    SecurityRequestContext.setUserId(ALICE.getName());
+
+    final FlowManager flowManager = appManager.getFlowManager(CrossNsDatasetAccessApp.FLOW_NAME);
+
+    testCrossNSSystemDatasetAccessWithAuthFlowlet(flowManager);
+    testCrossNSDatasetAccessWithAuthFlowlet(flowManager);
+
+    appManager.stopAll();
+  }
+
+  public void testCrossNSSystemDatasetAccessWithAuthFlowlet(final FlowManager flowManager) throws Exception {
+    addDatasetInstance(Id.Namespace.SYSTEM, "keyValueTable", "store");
+
+    // give bob write permission on the dataset
+    grantAndAssertSuccess(NamespaceId.SYSTEM.dataset("store"), BOB, EnumSet.of(Action.WRITE));
+
+    // switch to BOB
+    SecurityRequestContext.setUserId(BOB.getName());
+
+    Map<String, String> args = ImmutableMap.of(
+      CrossNsDatasetAccessApp.OUTPUT_DATASET_NS, NamespaceId.SYSTEM.getNamespace(),
+      CrossNsDatasetAccessApp.OUTPUT_DATASET_NAME, "store"
+    );
+
+    // But trying to run a flow as BOB will fail since this flow writes to a dataset in system namespace
+    flowManager.start(args);
+    // wait for flow to be running
+    Tasks.waitFor(true, new Callable<Boolean>() {
+      @Override
+      public Boolean call() throws Exception {
+        return flowManager.isRunning();
+      }
+    }, 120, TimeUnit.SECONDS);
+
+    // The above will be a runtime failure after the flow start since it will not be able to use the dataset in the
+    // system namespace. Since the failure will lead to no metrics being emitted we cannot actually check it tried
+    // processing or not. So stop the flow and check that the output dataset is empty
+    flowManager.stop();
+
+    verifyDatasetIsEmpty(NamespaceId.SYSTEM, "store");
+
+    // switch to back to ALICE
+    SecurityRequestContext.setUserId(ALICE.getName());
```

Comment:
why do you need to switch back to alice here? she has admin/write on NamespaceId.SYSTEM?